### PR TITLE
Feature: add a file manager system

### PR DESCRIPTION
**Pull request by MatthieuSLR9**
# Description

This is the first part of the file manager system. In it I create a popup to navigate through the files in a concise manner. For now there is a button on the overview to access this feature but in the future it will be accessed from the notes and folders to move them around. I modified slightly the ViewModel of the filesytem by adding a ney method in order for the state to not change when we query the subfolders of a folder. Furthermore, I fixed some problems when the current folder wasnt properly set in the view model.

# How Has This Been Tested?

Added UI tests to test the popup. The tests would go click one of the folders to make sur that we can navigate in the popup.
  

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged in downstream modules, no conflicts with main

### CODE REVIEW

File: app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt [hunk 1]
```diff
@@ -248,6 +249,37 @@ class OverviewTest {
...
+    composeTestRule.onNodeWithTag("showFileSystemButton").performClick()
+    composeTestRule.onNodeWithTag("FileSystemPopup").assertIsDisplayed()
+  }
+
+  @Test
+  fun navigateFileSystem() {
+    val userRootFoldersFlow =
+        listOf(
+            Folder(id = "8", name = "Root Folder 1", userId = "1", parentFolderId = "1"),
+            Folder(id = "9", name = "Root Folder 2", userId = "2", parentFolderId = "1"))
```
Comment (AvillegasH) [inline app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt]: For the second folder, the userId should be "1" to have it belong to user "1".

File: app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt [hunk 2]
```diff
@@ -248,6 +249,37 @@ class OverviewTest {
...
+      onSuccess(userRootFoldersFlow)
+    }
+    folderViewModel.getRootFoldersFromUid("1")
+
+    composeTestRule.onNodeWithTag("showFileSystemButton").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("showFileSystemButton").performClick()
+    composeTestRule.onNodeWithTag("FileSystemPopup").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("FileSystemPopupFolderChoiceBox8").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("FileSystemPopupFolderChoiceBox9").assertIsDisplayed()
+    // composeTestRule.onNodeWithTag("FileSystemPopupFolderChoiceText").assertIsDisplayed()
```
Comment (AvillegasH) [inline app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt]: Maybe consider removing the commented lines if you are not using them

File: app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt [hunk 3]
```diff
@@ -267,16 +299,6 @@ class OverviewTest {
 
   @Test
   fun dragAndDropNoteWorksCorrectly() {
-    `when`(noteRepository.getRootNotesFrom(eq("1"), any(), any())).then { invocation ->
```
Comment (AvillegasH) [inline app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt]: Why did you remove this test?

File: app/src/main/java/com/github/onlynotesswent/model/folder/FolderViewModel.kt [hunk 4]
```diff
@@ -67,6 +67,15 @@ class FolderViewModel(private val repository: FolderRepository) : ViewModel() {
   fun selectedFolder(folder: Folder) {
     _selectedFolder.value = folder
   }
+  /**
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/model/folder/FolderViewModel.kt:70]: Good addition of functions to handle folders and subfolders states. However, what do you think of modifying selected folder to accept a nullable argument so that we can directly set it to null by calling `selectedFolder(null)` instead of using clearSelectedFolder

File: app/src/main/java/com/github/onlynotesswent/model/folder/FolderViewModel.kt [hunk 5]
```diff
@@ -255,6 +264,15 @@ class FolderViewModel(private val repository: FolderRepository) : ViewModel() {
         onFailure = onFailure)
   }
 
+  fun getSubFoldersOfNoStateUpdate(
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/model/folder/FolderViewModel.kt:267]: Consider adding the corresponding documentation for this function

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt [hunk 6]
```diff
@@ -137,6 +149,108 @@ fun CreationDialog(
...
+ * Displays a popup dialog to browse and navigate the file system.
+ *
+ * This popup allows users to view folders in a hierarchical structure. Users can navigate into
+ * subfolders, view folder contents, and return to the root folder.
+ *
+ * @param onDismiss Callback invoked when the popup is dismissed.
+ * @param folderViewModel The ViewModel that provides the folder data.
+ */
+@Composable
+fun FileSystemPopup(onDismiss: () -> Unit, folderViewModel: FolderViewModel) {
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:162]: Out of scope for this PR, but in a future PR we should make this pop up more general so that we can reuse for decks and flashcards

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt [hunk 7]
```diff
@@ -137,6 +149,108 @@ fun CreationDialog(
...
+                      color = MaterialTheme.colorScheme.onPrimary)
+                }
+
+            Column(
+                modifier =
+                    Modifier.weight(1f)
+                        .padding(start = 16.dp, end = 16.dp, top = 8.dp)
+                        .verticalScroll(rememberScrollState()),
+                verticalArrangement = Arrangement.spacedBy(4.dp)) {
+                  if (selectedFolder == null) {
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:197]: To improve readability, creating a Boolean Mutable State (called root for instance) containing whether selected folder is null could be added.

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt [hunk 8]
```diff
@@ -137,6 +149,108 @@ fun CreationDialog(
...
+                    userRootFolders.value.forEach { folder ->
+                      Box(
+                          modifier =
+                              Modifier.testTag("FileSystemPopupFolderChoiceBox" + folder.id)
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:201]: Good idea to have independent test tags for different folders

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt [hunk 9]
```diff
@@ -137,6 +149,108 @@ fun CreationDialog(
...
+                                  .padding(12.dp)) {
+                            Text(
+                                modifier = Modifier.testTag("FileSystemPopupFolderChoiceText"),
+                                text = folder.name,
+                                style = MaterialTheme.typography.bodyMedium,
+                                color = MaterialTheme.colorScheme.onSurface)
+                          }
+                    }
+                  } else {
+                    folderSubFolders.forEach { subFolder ->
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:221]: Maybe since both statements in the if and else are quite similar, extracting them as a separated function could be done

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt [hunk 10]
```diff
@@ -137,6 +149,108 @@ fun CreationDialog(
...
+                                style = MaterialTheme.typography.bodyMedium,
+                                color = MaterialTheme.colorScheme.onSurface)
+                          }
+                    }
+                  }
+                }
+
+            Box(modifier = Modifier.fillMaxWidth().padding(8.dp)) {
+              ElevatedButton(
+                  onClick = { selectedFolder = null }, modifier = Modifier.fillMaxWidth()) {
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:247]: Consider using the added function in the view model to set the state to null

File: app/src/main/java/com/github/onlynotesswent/ui/overview/FolderContent.kt [hunk 11]
```diff
@@ -93,14 +94,17 @@ fun FolderContentScreen(
 
   var updatedName by remember { mutableStateOf(folder.value!!.name) }
   var expandedFolder by remember { mutableStateOf(false) }
-
+  var showFileSystemPopup by remember {
+    mutableStateOf(true)
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/overview/FolderContent.kt:98]: Shouldnt the default value of showFileSystemPopUp be false?

File: app/src/main/java/com/github/onlynotesswent/ui/overview/FolderContent.kt [hunk 12]
```diff
@@ -138,6 +142,11 @@ fun FolderContentScreen(
               context = context,
               folder = folder.value)
         }) { paddingValues ->
+          Button(onClick = { showFileSystemPopup = true }) { Text("FileSystem Overview") }
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/overview/FolderContent.kt:145]: Consider adding the Text string to the strings.xml file

File: app/src/main/java/com/github/onlynotesswent/ui/overview/FolderContent.kt [hunk 13]
```diff
@@ -138,6 +142,11 @@ fun FolderContentScreen(
...
+          Text(
+              text = "Selected Folder: ${folderViewModel.selectedFolder.value?.name ?: "None"}",
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/overview/FolderContent.kt:147]: Same as before for selected folder and none

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt [hunk 14]
```diff
@@ -137,6 +149,108 @@ fun CreationDialog(
...
+                    Text("Back to Root Folder")
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:248]: Consider adding the string to strings.xml file

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt [hunk 15]
```diff
@@ -137,6 +149,108 @@ fun CreationDialog(
...
+                    shape = RoundedCornerShape(12.dp))) {
+          Column(modifier = Modifier.fillMaxWidth().fillMaxHeight()) {
+            Box(
+                modifier =
+                    Modifier.fillMaxWidth()
+                        .background(MaterialTheme.colorScheme.primary)
+                        .padding(vertical = 12.dp, horizontal = 16.dp)) {
+                  Text(
+                      text =
+                          if (selectedFolder == null) "Folders in: root"
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:185]: Consider adding folders in: to strings.xml file

**Review by AvillegasH — CHANGES_REQUESTED** [review]
Overall, very good code and added features. Documentation and testing were also of good quality. Other than that, consider changing some of the comments I added and think about some of the suggestions I made. The UI already looks good, but with the second PR Im sure it will look even better.

File: app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt (hunk 1 above)
Comment (MatthieuSLR9) [reply app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt]: Yes I will change that, I wrote random user numbers as I do not really care about them in my tests but it is more coherent to have the same user number.

File: app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt (hunk 2 above)
Comment (MatthieuSLR9) [reply app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt]: Yes this was already removed in my latest push.

File: app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt (hunk 3 above)
Comment (MatthieuSLR9) [reply app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt]: I removed it by error, the latest push should reput it.

File: app/src/main/java/com/github/onlynotesswent/model/folder/FolderViewModel.kt (hunk 5 above)
Comment (MatthieuSLR9) [reply app/src/main/java/com/github/onlynotesswent/model/folder/FolderViewModel.kt:267]: Yes good point I will do that.

File: app/src/main/java/com/github/onlynotesswent/model/folder/FolderViewModel.kt (hunk 4 above)
Comment (MatthieuSLR9) [reply app/src/main/java/com/github/onlynotesswent/model/folder/FolderViewModel.kt:70]: I think it is cleaner to have a clearFunction. It would seem a bit weird I feel like to call selectedFolder(null). When reading the code I feel like having a explicit clearSelectedFolder method to be better.

File: app/src/main/java/com/github/onlynotesswent/ui/overview/FolderContent.kt (hunk 12 above)
Comment (MatthieuSLR9) [reply app/src/main/java/com/github/onlynotesswent/ui/overview/FolderContent.kt:145]: Yes good point

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt (hunk 9 above)
Comment (MatthieuSLR9) [reply app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:221]: I left it as is as in the second PR I think I'll have a different content depending if we are at Root

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt (hunk 10 above)
Comment (MatthieuSLR9) [reply app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:247]: I thought it was better to do this like that as here we are referencing our local selectedFolder and do not need to go through the viewModel.

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt (hunk 6 above)
Comment (MatthieuSLR9) [reply app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:162]: Yes that is the goal of my task for this week.

File: app/src/main/java/com/github/onlynotesswent/ui/overview/FolderContent.kt (hunk 11 above)
Comment (MatthieuSLR9) [reply app/src/main/java/com/github/onlynotesswent/ui/overview/FolderContent.kt]: Yes, I actually scrapped the button from the FolderContent as for now our button only is used temporarily to see my features work but the actual place the popup should be displayed is from the drop down menus of notes and an option for folders which will be done later.

File: app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt [hunk 16]
```diff
@@ -318,4 +319,47 @@ class OverviewTest {
           up()
         }
   }
+
```
Comment (AvillegasH) [inline app/src/androidTest/java/com/github/onlynotesswent/ui/overview/OverviewTest.kt:322]: Nice changes to complete the tests

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt (hunk 9 above)
Comment (AvillegasH) [reply app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:222]: Understood, no problem

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt (hunk 10 above)
Comment (AvillegasH) [reply app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:248]: True, even though I think changes in the general state should be observed in the local one

File: app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt [hunk 17]
```diff
@@ -137,6 +149,110 @@ fun CreationDialog(
...
+                                color = MaterialTheme.colorScheme.onSurface)
+                          }
+                    }
+                  }
+                }
+
+            Box(modifier = Modifier.fillMaxWidth().padding(8.dp)) {
+              ElevatedButton(
+                  onClick = { selectedFolder = null }, modifier = Modifier.fillMaxWidth()) {
+                    Text(stringResource(R.string.file_system_back_to_root))
```
Comment (AvillegasH) [inline app/src/main/java/com/github/onlynotesswent/ui/common/Popups.kt:249]: Good use of stringResource

**Review by AvillegasH — APPROVED** [review]
